File: Web Survey/resources/js/recommender/voting_rules/AbstractVotingRule.js
export default class AbstractVotingRule {

    constructor(k=-1) {
        this.k = k;
        this.ruleName = this.constructor.getRuleName();
    }

    static getRuleName() {
        return "Abstract Voting Rule"
    }

    static get idName() {
        return this.getRuleName().replace(/ /g, "").replace("'", "");
    }

    getDescription() {
        return "Abstract class of the voting rules"
    }

    getUsers(data) {
        return Object.keys(Object.values(data)[0]);
    }

    sort(sortArray) {
        sortArray.sort(function(x, y) {
            if (x[1] < y[1]) {
                return -1;
            }
            if (x[1] > y[1]) {
                return 1;
            }
            return 0;
        }).reverse();
    }

    votingRule(song, songRatings) {
        throw new Error("Calling abstract method");
    }

    *calculateVotes(songRatings) {

        let votingData = [];
        for(const song in songRatings) {
            if(songRatings.hasOwnProperty(song)) {
                votingData.push([song, this.votingRule(song, songRatings[song])]);
            }
        }

        //return sorted(result.items(), key=lambda v: v[1], reverse=True)
        this.sort(votingData);

        let length;
        if(this.k === -1) {
            length = votingData.length;
        } else {
            length = this.k;
        }

        for(const i in [...Array(length).keys()]) {
            yield votingData[i];
        }
    }

}

File: Experiment Processing/recommender/voting_rules/abstract_voting_rule.py
from abc import ABC, abstractmethod


class AbstractVotingRule(ABC):
    """
    Abstract class for the different aggregation strategies (called voting rules at an early stage).
    The voting rules can be used by first initialising with a list of user ids.
    Then, the calculate_votes function can be called to retrieve a generator which will yield all items in descending
    order based on the calculated score.

    Note: only PWS, Fairness, and Least Misery were used for the experiments, but more strategies were implemented.
    """

    def __init__(self, users):
        self.users = users

    @abstractmethod
    def get_name(self):
        pass

    @abstractmethod
    def voting_rule(self, song, song_ratings):
        pass

    def calculate_votes(self, data):
        result = {}

        for song in data:
            result[song] = self.voting_rule(song, data[song])

        sorted_list = sorted(result.items(), key=lambda v: v[1], reverse=True)
        for track in sorted_list:
            yield track
